Style Engine: Add a WP_Style_Engine_Processor object

aristath opened the pull request:
Following discussions in other tickets for the Style Engine, this PR adds a ~~`Renderer`~~ `Processor` to the style engine.

## Usage:

A complete example that can be used to test this:
```php
// Get the store.
$store = WP_Style_Engine_CSS_Rules_Store_Gutenberg::get_store( 'global-styles' );
// Add rules & declarations.
$store->add_rule( '.wp-block-button a:hover' )
	->add_declarations( [
		'color'            => 'var(--block-button-link-hover-color)',
		'outline'          => '1px solid var(--block-button-link-hover-color)',
		'background-color' => 'black',
	] )->add_declarations( [
		'background-color' => '#000',
	] );

$store->add_rule( '.foo' )->add_declarations( [
	'color'     => 'red',
	'font-size' => '1em',
] );
$store->add_rule( '.bar' )->add_declarations( [
	'font-size' => '1em',
	'color'     => 'red',
] );

// Init the Renderer object
$processor = new WP_Style_Engine_Processor_Gutenberg( $store );
// Get the CSS.
$css = $processor->get_css();
```

With the above code, the `$css` at the end is `.wp-block-button a:hover {color: var(--block-button-link-hover-color); background-color: #000;}.foo,.bar {color: red; font-size: 1em;}`

Please note the following in the final CSS:
* In the `.wp-block-button a:hover` selector we're adding `background-color: black`, and then `background-color: #000`. In the final CSS the latter overrides the former so there are no duplicate entries for CSS properties under the same selector. This was part of previous PRs so not specifically a part of the Renderer, but worth highlighting since it is a big improvement over what we have now with strings concatenations.
* We're adding separate rules for `.foo` and `.bar`, but they have the same declarations so in the final CSS they get combined to `.foo,.bar` (reusing the concept from https://github.com/WordPress/gutenberg/pull/41896)

aristath committed:
Add a WP_Style_Engine_Renderer object

aristath committed:
Better use of the engine objects

andrewserong commented on packages/style-engine/class-wp-style-engine-renderer.php:
```diff
@@ -0,0 +1,92 @@
+ *
+ * @access private
+ */
+class WP_Style_Engine_Renderer {
```
In terms of naming, I was wondering what you and @ramonjd think about potentially calling this an `Optimizer` instead of a renderer. Something that's been on my mind lately is how the naming of the style engine and its component parts can set expectations of what it will (or might be able to do) and `Renderer` suggests to me that it might also handle other things than optimizing CSS and outputting CSS rules.

I don't feel too strongly about it, though, if you think Renderer is a better fit!

ramonjd replied:
That sounds fair. This processing is the last step before we publish/render the 'optimized' styles to the page.

There's probably a semantic argument for any word here 😄 , but I agree I think we could choose a different term just to be clear... especially if, in the context of the style engine, we use 'render' to mean something like page/browser rendering. 👍 

`Optimizer` sounds fine, or `Processor/[insert synonym]`, given that this class appears to be a good place to add further pre-render tasks.

aristath replied:
It depends on how we envision the Renderer object...
This PR merely adds the 1st step to what I hope the Rendered will become.
Ideally, the Renderer with combine the CSS from all rules in a store (done in this PR), optimize the CSS (somewhat done in this PR), and then enqueue the generated styles (pending)
Enqueueing the CSS is not handled in this PR because we'll need to thing that part of the object some more and figure out the details for all cases. Since this is about inline styles, they will most likely be added using `wp_add_inline_styles`, which means they'll need to be "attached" to a pre-existing stylesheet. But not all stores will use the same stylesheet... So we need to think how to implement that part some more before committing something.
With the above in mind, I called the object `Renderer`... the CSS optimizing might be a relatively big part of the code here, but is just a small part of the final envisioned implementation.

We could however call this one `Processor`, and then create a separate `Renderer` if and when the need for it arises 🤔 

Thoughts?

aristath replied:
Actually, the more I think of this, the more `Processor` makes sense... 
We'll be processing and generating and final CSS. How that CSS is used/enqueued can be left to the individual implementations.
I'll go ahead and rename the object 👍

andrewserong reviewed (commented):
I like the direction this is taking, too! Just added a small comment to discuss naming.

I was also wondering if there are performance considerations for us to consider with doing an MD5 check on each declaration? For example, if we're `json_encode`ing to a string, can we do a string comparison on the JSON instead of the additional step of using an `MD5` check, or is there a benefit from the extra step? PHP performance is definitely not my strong suit, so I wasn't sure on balance which is faster (comparing potentially _really_ long strings, or doing an MD5 check first, and then comparing short strings) 🤔

ramonjd committed:
Added tests

ramonjd commented:
> PHP performance is definitely not my strong suit, so I wasn't sure on balance which is faster (comparing potentially really long strings, or doing an MD5 check first, and then comparing short strings

This is a great point, and it's something I've had in the back of my head as we add more overhead. 

I've added a task to the [project board](https://github.com/orgs/WordPress/projects/19/views/1) to integrate some performance/speed tests.

ramonjd commented:
I added some tests, hope you don't mind! I like to see the output under various conditions. 

It's looking great! Especially the combine/merging functionality.

ramonjd committed:
Updating method name to add_declarations after rebase

aristath committed:
no need to use md5

aristath commented:
> I like the direction this is taking, too! Just added a small comment to discuss naming.
> 
> I was also wondering if there are performance considerations for us to consider with doing an MD5 check on each declaration? For example, if we're `json_encode`ing to a string, can we do a string comparison on the JSON instead of the additional step of using an `MD5` check, or is there a benefit from the extra step? PHP performance is definitely not my strong suit, so I wasn't sure on balance which is faster (comparing potentially _really_ long strings, or doing an MD5 check first, and then comparing short strings) 🤔

Good point. I removed the extra `md5` step and it now compares the JSON strings 👍

aristath committed:
No need for an extra method

aristath committed:
renaming stuff

aristath committed:
micro-optimization

aristath committed:
rename Renderer to Processor

aristath committed:
:facepalm: forgot to rename files on inclusion

aristath committed:
some inline comments

aristath commented:
I believe this is now ready for a final review... 
Architecturally it makes sense, performance-wise it's fast, it's simple since the only exposed method in the object is `get_css()` so it leaves a lot of room for future implementations, and all tests pass (huge thanks @ramonjd!)

Once this gets in we can start refactoring global-styles and block-supports to use the Style Engine 🎉

andrewserong reviewed (approved):
Thanks for all the follow-up @aristath (and the tests @ramonjd), this is looking good to go to me now, and nice that all the logic is happening in the Processor 👍

Looks good to merge to me, and we can always tweak in follow-ups if need be, too. 🚀
